platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +465,242 @@ public DataFileReader endWriting() throws IOException {
+                // Create a "read bucket" task
+                final ReadUpdateBucketTask readBucketTask = new ReadUpdateBucketTask(getPool(), bucketIndex, bucketMap);
+                // Execute it right away
+                readBucketTask.send();
+            }
+            if (it.hasNext()) {
+                // There are more buckets to process. Let the next submit task run. One of the next task's
+                // dependencies is set here, the other one is set in the end of StoreBucketTask
+                nextSubmitTask.send();
+            }
+            return true;
+        }
+    }
+
+    /**
+     * A task to read a bucket identified by the given idex from disk and apply a list of
+     * updates to the keys to it. The task has no dependencies, it's executed right after
+     * creation.
+     */
+    private class ReadUpdateBucketTask extends AbstractTask {
+
+        // Bucket index
+        private final int bucketIndex;
+
+        // List of updates to apply to the bucket
+        private final BucketMutation keyUpdates;
+
+        ReadUpdateBucketTask(final ForkJoinPool pool, final int bucketIndex, final BucketMutation keyUpdates) {
+            super(pool, 0);
+            this.bucketIndex = bucketIndex;
+            this.keyUpdates = keyUpdates;
+        }
+
+        private void createAndScheduleStoreTask(final Bucket bucket) {
+            // Create a subsequent "store bucket" task for the bucket
+            final StoreBucketTask storeTask = new StoreBucketTask(getPool(), bucket);
+            // The last created "store bucket" task. storeTask above will be set as an
+            // output dependency for that task to make sure tasks are running only one at
+            // a time. See StoreBucketTask for details
+            final StoreBucketTask prevTask = lastStoreTask.getAndSet(storeTask);
+            if (prevTask != null) {
+                // This will trigger prevTask execution as soon as its prev task is complete
+                prevTask.setNext(storeTask);
             } else {
-                bucket = bucketPool.getBucket();
-                bucket.readFrom(bucketData);
+                // The first task: no dependency on the prev task, can be executed rightaway
+                storeTask.send();
+            }
+            if (storeBucketTasksCreated.incrementAndGet() == numBuckets.get()) {
+                // The last task: no dependency on the next task, can be executed as soon as
+                // its prev task is complete, no need to wait until the next task dependency
+                // is set
+                lastStoreTask.get().setNext(notifyTaskRef.get());
+            }
+        }
+
+        @Override
+        protected boolean exec() {
+            try {
+                BufferedData bucketData =
+                        fileCollection.readDataItemUsingIndex(bucketIndexToBucketLocation, bucketIndex);
+                // The bucket will be closed by WriteBucketTask
```

Comment:
s/WriteBucketTask/StoreBucketTask/

Reply:
Noted. I'll fix this comment in some future fixes.